design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,389 @@
+used with `kubectl` to help automate this manual configuration.
+
+### Implementation Details/Notes/Constraints
+
+One of the primary drivers of his design was that each vendor uses different
+names and values for similar BIOS Configuration settings. This design does not
+rely on vendor specific checks but instead uses common mechanisms to ensure
+the BIOS Settings are valid:
+
+- copying the names and values from `settingsData` to
+  `vendorSettings` ensures initial values are correct
+- when updating the values in `vendorSettings`, the user can use the
+  limitations in the `settingsData`, e.g. minimum and maximum values
+- before writing to the Ironic clean-steps API, the BMO runs validation checks
+  against the `settingsData` to ensure updated values are valid
+
+In order to interoperate with the individual BIOS Configuration set
+approach the BMO must ensure that the same settings don't get added twice
+to the Ironic clean-steps API. In addition, as each clean step requires a
+reboot the number of clean steps that are added must be managed.
+
+An example of the proposed CRD definition for the BIOS Settings is as follows:
+
+```yaml
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.4.1
+  creationTimestamp: null
+  name: hostfirmwaresettings.metal3.io
+spec:
+  group: metal3.io
+  names:
+    kind: HostFirmwareSettings
+    listKind: HostFirmwareSettingsList
+    plural: hostfirmwaresettings
+    singular: hostfirmwaresettings
+  scope: Namespaced
+  versions:
+  - name: v1alpha1
+    schema:
+      openAPIV3Schema:
+        description: HostFirmwareSettings is Schema for hostfirmwaresettings
+        properties:
+          apiVersion:
+            description: 'APIVersion defines the versioned schema'
+            type: string
+          kind:
+            description: 'Kind is a string value representing the resource'
+            type: string
+          metadata:
+            type: object
+          spec:
+            description: HostFirmwareSettingsSpec defines the desired state
+            properties:
+              vendorSettings:
+                items:
+                  description: describes a setting whose value can be changed
+                  properties:
+                    name:
+                      description: Identifier for the Firmware setting.
+                      type: string
+                    value:
+                      description: Value of the Firmware setting.
+                      type: string
+                  required:
+                  - name
+                  - value
+                  type: object
+                type: array
+            type: object
+          status:
+            description: defines the observed state of HostFirmwareSettings
+            properties:
+              settingsData:
+                items:
+                  description: describes fields read from Ironic
```

Comment:
Minor: _"describes the current host firmware settings along with their schema_" (I think the reference to Ironic could also be safely removed, as it is not relevant in this context)

Reply:
Agree, done.